Port overlays documentation (#171)
microsoft/vcpkg-docs: vcpkg/concepts/overlay-ports.md

---

> title: Port overlays

Reviewer: ```suggestion
title: Overlay ports
```

description: This article describes overlay ports in vcpkg. Overlay ports are used to force a specific port definition to be used by vcpkg during the package installation process.
author: data-queue
ms.author: danshaw2
ms.date: 9/20/2023
ms.prod: vcpkg
ms.topic: concept-article
---

# Overlay ports

> An overlay port can act as a drop-in replacement for an existing port. An overlay port can also add a new port that is otherwise not available in a [registry](../maintainers/registries.md). While [resolving package names](../users/registries.md#package-name-resolution), overlay ports take priority, forcing overlay ports to be selected.

Reviewer: ```suggestion
An overlay port is a port that can act as a drop-in replacement for an existing port. An overlay port can also be a new port that is otherwise not available in a [registry](../maintainers/registries.md). While [resolving package names](../users/registries.md#package-name-resolution), overlay ports take priority, forcing overlay ports to be selected.
```

> An overlay port is a port that can act as a drop-in replacement for an existing port. An overlay port can also be a new port that is otherwise not available in a [registry](../maintainers/registries.md). While [resolving package names](../users/registries.md#package-name-resolution), overlay ports take priority, forcing overlay ports to be selected.

Reviewer: ```suggestion
Usually, vcpkg ports are obtained from [registries](../users/registries.md). It is very likely that most of the ports you install come from the official vcpkg registry at <https://github.com/Microsoft/vcpkg>. vcpkg lets you install ports available to you via the file system, we call these ports, overlay ports.

An overlay port can act as a drop-in replacement for an existing port or as a new port that is otherwise not available in a [registry](../maintainers/registries.md). While [resolving package names](../users/registries.md#package-name-resolution), overlay ports take priority.
```

> Naming conflicts can arise when multiple overlays introduce the same port. For instance, overlay directories `[dirA, dirB]` may introduce a port with the same name: `dirA/zlib` and `dirB/zlib`. vcpkg resolves `zlib` in order of which directory is specified first. This example selects `dirA/zlib` as the `zlib` port. If overlay port options are specified in multiple places, overlay resolution priority follows first from the command line, then the manifest, and finally, the environment.

Reviewer: ```suggestion
Overlay ports are evaluated in the following order:

* Overlay locations specified in the command-line via [`--overlay-ports`](../commands/common-options.md#overlay-ports).
* Overlay locations specified in a `vcpkg-configuration.json` file via [`overlay-ports`](../reference/vcpkg-configuration-json.md).
* Overlay locations specified by the [`VCPKG_OVERLAY_PORTS`](../users/config-environment.md#vcpkg_overlay_ports) environment variable.

When resolving port names, the first location that contains a matching overlay port is selected.
```

## Using an overlay port

> A directory refers to a set of overlay ports and can be specified as follows: 

Reviewer 2: ```suggestion
A directory can represent a set of overlay ports and can be specified in one of two ways:
```

- Single overlay port: `<directory>/sqlite3` refers to a single port
- Directory of overlay ports: `<directory>` refers to a directory of ports
A valid port must contain both `vcpkg.json` and `portfile.cmake`.

You can add an overlay port in several ways:
* Command-line: Add one or multiple `--overlay-ports=<directory>` options to your vcpkg command
* [Manifest](../reference/vcpkg-configuration-json.md#overlay-ports): Populate the `"overlay-ports"` array in `vcpkg-configuration.json`
* [Environmental variable](../users/config-environment.md#vcpkg_overlay_ports): Set `VCPKG_OVERLAY_PORTS` to a list of directory paths

### Example: Overlay Ports Example

Given this directory structure:

:::image type="complex" source="../resources/ports-overlay-example.png" alt-text="Example with multiple overlay port directories":::
Overlay directory named team-ports contains ports sqlite3, rapidjson and curl. Overlay directory named my-ports contains ports sqlite3 and rapidjson. The vcpkg directory contains the default catalog.
:::image-end:::

> Run:
> ```bash
> vcpkg install sqlite3 --overlay-ports=my-ports --overlay-ports=team-ports
> ```
> To install:
> - `sqlite3` from `my-ports`

Reviewer 2: ```suggestion
To install `sqlite3` from `my-ports`, run:

```bash
vcpkg install sqlite3 --overlay-ports=my-ports --overlay-ports=team-ports
```

```

Run:
```

> vcpkg/vcpkg.exe install sqlite3 rapidjson curl 

Reviewer 2: ```suggestion
vcpkg install sqlite3 rapidjson curl 
```

Just to be consistent with how you referred to vcpkg above

> Run:
> ```
> vcpkg/vcpkg.exe install sqlite3 rapidjson curl 
>     --overlay-ports=my-ports/rapidjson 
>     --overlay-ports=vcpkg/ports/curl
>     --overlay-ports=team-ports
> ```
> To install:
> - `sqlite3` from `team-ports`
> - `rapidjson` from `my-ports`
> - `curl` from `vcpkg/ports

Reviewer 2: ```suggestion
To install:
- `sqlite3` from `team-ports`
- `rapidjson` from `my-ports`
- `curl` from `vcpkg/ports`

Use the following command:

```bash
vcpkg install sqlite3 rapidjson curl 
    --overlay-ports=my-ports/rapidjson 
    --overlay-ports=vcpkg/ports/curl
    --overlay-ports=team-ports
```.
```

Reviewer 2: This way, readers will have an understanding of the expected result before executing the command.

Author: I don't think we expect users to run the command, it serves more as an example. I've kept the changes as before.
